odoo/addons/base/models/ir_attachment.py

```diff
@@ -50,21 +90,26 @@ def _storage(self):
     def _filestore(self):
         return config.filestore(self._cr.dbname)
 
+    @api.model
+    def storage_locations(self):
+        return ['db', 'file']
+
     @api.model
     def force_storage(self):
-        """Force all attachments to be stored in the currently configured storage"""
         if not self.env.is_admin():
             raise AccessError(_('Only administrators can execute this action.'))
-
-        # domain to retrieve the attachments to migrate
-        domain = {
-            'db': [('store_fname', '!=', False)],
-            'file': [('db_datas', '!=', False)],
-        }[self._storage()]
-
-        for attach in self.search(domain):
-            attach.write({'datas': attach.datas})
-        return True
+        # Migrate only binary attachments and make sure to miragte those that belong 
+        # to a field as well as attachments that are not assigned to a field. 
+        self.search(expression.AND([self._get_storage_domain(self._storage()), [
+            '&', ('type', '=', 'binary'), '|', ('res_field', '=', False), ('res_field', '!=', False) 
```

Comment:
This change is necessary to migrate attachment which fields are assigned as well. Other the `_search` method would filter all of those attachments.

```python
if not any(arg[0] in ('id', 'res_field') for arg in args):
    args.insert(0, ('res_field', '=', False))
```

Checking for type `binary` is only a small optimization.